### ISSUE TITLE
[RFC] Ray Serve model multiplexing support

### ISSUE THREAD

**Issue by sihanwang41** [ray-project/ray#33253]
TL;DR - This RFC proposes a new API to serve multiple models on a single Ray Serve deployment without splitting the models into separate deployments or Serve applications.

## Problem Statement

Recently, the Serve team has seen growing interest in serving many independent models. In Ray 2.3 we added multi-app support, so you could run multiple, independent Serve applications on a single Ray cluster. Multi-app support lets Serve apps share resources at the cluster level, but there are some use cases better suited to sharing resources at the deployment level.

For example, a user may wish to run a Recommender deployment that serves personalized recommendations to their customers. Each end customer may have a personalized model that’s fine-tuned for their needs. For this use case, the user may have hundreds of recommendation models that should all be served by Recommender deployment replicas, rather than getting split into individual deployments for each model.

For this use case, Serve could provide an API that lets deployment replicas transparently load model weights from a set of models, route traffic to replicas that contain the requested model, and evict model weights whenever new models are needed.

In short, Serve deployments would offer model multiplexing.

## Proposal

Introduce a new deployment decorator, @serve.multiplex, to Ray Serve. You can pass the argument max_models_per_replicas to the decorator to control how many models can be loaded within each replica.

```
@serve.deployment(num_replicas=6)
class UserDeployment:
    def __init__(self, model_id: str):
        self._model = load_model(model_id)
        logger.info(f"loading model from {model_path}")

        @serve.multiplexed(max_num_models_per_replica=5)
        async def load_model(self, model_id: str) -> Any:
            # Load model with the given tag
            # You can use any model loading library here
            # and return the loaded model. load_from_s3 is
            # a placeholder function.
            return load_from_s3(model_id)

    def __call__(self, request: starlette.requests.Request):
            # Get the model_id from the request context.
            model_id = serve.get_multiplexed_model_id()
            # Load the model for the requested model_id.
            # If the model is already cached locally,
            # this will just be a dictionary lookup.
            model = await self.load_model(model_id)
            return model(request)
```

UserDeployment will have 6 replicas, each running at most 3 models.

To run this code
```
serve.run(UserDeployment.bind(), route_prefix=”/app”)
```

To query the endpoint, Ray Serve will use the model_id as tag to route the traffic
```
resp = requests.get(
    "http://127.0.0.1:8000/app", headers={"ray_serve_request_routing_tag": "1"}
)
```

When there’s no space left to load a new model in the cluster, Serve will use an LRU policy to evict existing models from an arbitrary replica and load the new model.

Autoscaling will still happen on the replica level. There’s no per-model autoscaling with this proposal. Replicas will be scaled up or down based on the QPS  and request queue metrics. 

There could potentially be more advanced options in the @serve.multiplex decorator, such as memory size limit, qps limit, etc. In this proposal, we start with the simplest one: setting the number of models per replica.

**Comment by VishDev12** [comment]
Hello! This seems like an interesting feature to have! A couple of questions:

1. Please correct me if I'm mistaken, but I've noticed that Ray Serve was positioned as a generic tool not specific to ML. So does having the name: `model` introduced here affect that?
2. `UserDeployment.bind()` initializes the Deployment, but when exactly is `__init__` triggered? Since I see it requires a `model_id`. And what is the interface if a new sub-model needs to be added?
3. What does `control how many models can be loaded within each replica` mean? If a replica is online, the assumption is that the `__init__` has run and the model was loaded. And in your example, I see a single `model_id` passed to `__init__`. How can one replica run 3 `__init__`s?  Is my mistake equating a single replica with a single instance of the `UserDeployment` class?

Barring the concept of `max_models_per_replicas` which I think I failed to understand, I think this concept is nice! Because we can have one deployment of type: `CustomNetworkA` in the app, for example, and that deployment will have 10 replicas, each with its own set of weights accessible through custom endpoints. So it's saving us the trouble of defining 10 separate deployments.

But my worry is this, let's say one out of those models gets a lot of traffic and needs to be scaled independently. Aren't we back to the scaling issue that deployments solved?

How does this proposed solution equate to looping over a list and creating deployments with a custom route for each element in the list?

**Comment by JoshKarpel** [comment]
> How does this proposed solution equate to looping over a list and creating deployments with a custom route for each element in the list?

+1, this seems to be the key comparison that needs to be addressed. The existing model of separate deployments provides clear behavior around routing, initialization, getting handles to the deployment, etc., and is not overly-specialized for ML use cases. It seems like the area where this RFC could provide an improvement is in resource (CPU, memory, network) efficiency, but that would presumably have a lot to do with the implementation details.

**Comment by blublinsky** [comment]
Another question is how this is going to support Graphs, which is a great addition to serve

**Comment by sihanwang41** [comment]
Hi @VishDev12 , (sorry for my late reply!)

> 1. Please correct me if I'm mistaken, but I've noticed that Ray Serve was positioned as a generic tool not specific to ML. So does having the name: `model` introduced in `max_models_per_replicas` affect that? Is the concept of a model here something generic?

In this example, model_id is an example in this proposal, user can use different attribute, internally we will use the specific attribute assigned to `tag` which will be used for routing.

> 2. `UserDeployment.bind()` initializes the Deployment, but when exactly is `__init__` triggered? Since I see it requires a `model_id`. And what is the interface if a new sub-model needs to be added?

In this proposal, __init__ will happen when there is new model to be loaded into a replica. The `multiplex` decorator (for implementation, we will have an internal deployment class) will hold the `UserDeployment`.  

> 3. What does `control how many models can be loaded within each replica` mean? If a replica is online, the assumption is that the `__init__` has run and the model was loaded. And in your example, I see a single `model_id` passed to `__init__`. How can one replica run 3 `__init__`s in the `max_models_per_replicas=3` example you've provided? Is my mistake equating a single replica with a single instance of the `UserDeployment` class?

Same as 2), the __init__ will be run 3 times in this case.

> 
> Barring the concept of `max_models_per_replicas` which I think I failed to understand, I think this concept is nice! Because we can have one deployment of type: `CustomNetworkA` in the app, for example, and that deployment will have 10 replicas, each with its own set of weights accessible through custom endpoints. So it's saving us the trouble of defining 10 separate deployments.

That is the case we want to support! You don't need deploy 10 separate deployments by using multiplex decorator.

> 
> But my worry is this, let's say one out of those models gets a lot of traffic and needs to be scaled independently. Aren't we back to the scaling issue that deployments solved?
In the starting point of multiplex support, we are still going for by replica autoscaling. E.g. There is one model having large volume traffic, if it is exceed the `target_num_ongoing_requests_per_replica`, there will be new replicas provisioned. 
> 
> How does this proposed solution equate to looping over a list and creating deployments with a custom route for each element in the list?

When model size is small, user can combine multiple models into one replica. With [multi applications support ](https://github.com/ray-project/ray/issues/31072) in 2.x API, we are able to support multiple separate deployments + multiplexing in a hybrid approach.

**Comment by sihanwang41** [comment]
> Another question is how this is going to support Graphs, which is a great addition to serve
There is no special change on the graph support. you can have the actor handle of subsequent node, so that multiple models in one replica can use same subsequent node to continue the graph inference.

**Comment by tsinggggg** [comment]
how does this works with dynamic batching? with multiplexing, the requests in a batch could belong to different model id, then they couldn't effectively be batched together

**Comment by tsinggggg** [comment]
under multiplexing, for a single multiplex model id, if it's loaded in 2 replicas, do they share any resources in anyway? or it simply takes twice the memory needed to load the model once
